[SPARK-15616][SQL] Add optimizer rule PruneHiveTablePartitions

fuwhu opened the pull request:
### What changes were proposed in this pull request?
Add optimizer rule PruneHiveTablePartitions pruning hive table partitions based on filters on partition columns.
Doing so, the total size of pruned partitions may be small enough for broadcast join in JoinSelection strategy.

### Why are the changes needed?
In JoinSelection strategy, spark use the "plan.stats.sizeInBytes" to decide whether the plan is suitable for broadcast join.
Currently, "plan.stats.sizeInBytes" does not take "pruned partitions" into account, so it may miss some broadcast join and take sort-merge join instead, which will definitely impact join performance.
This PR aim at taking "pruned partitions" into account for hive table in "plan.stats.sizeInBytes" and then improve performance by using broadcast join if possible.

### Does this PR introduce any user-facing change?
no

### How was this patch tested?
Added unit tests.

This is based on #25919, credits should go to @lianhuiwang and @advancedxy.

fuwhu committed:
Add optimizer rule PruneHiveTablePartitions pruning hive table partitions based on filters on partition columns.
Doing so, the total size of pruned partitions may be small enough for broadcast join in JoinSelection strategy.

fuwhu commented:
@wangyum @cloud-fan 
Could you please help review ?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #115119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115119/testReport)** for PR 26805 at commit [`0b60978`](https://github.com/apache/spark/commit/0b609786fedc7d89f12bedf9e48c7616b1882c69).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19932/
Test PASSed.

SparkQA commented:
**[Test build #115119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115119/testReport)** for PR 26805 at commit [`0b60978`](https://github.com/apache/spark/commit/0b609786fedc7d89f12bedf9e48c7616b1882c69).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PruneHiveTablePartitions(session: SparkSession)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115119/
Test PASSed.

fuwhu commented:
gently cc: @cloud-fan @maropu

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20726/
Test PASSed.

SparkQA commented:
**[Test build #115934 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115934/testReport)** for PR 26805 at commit [`4e1aba9`](https://github.com/apache/spark/commit/4e1aba9c33dceed751db403f303e4bdb0c6e7cf6).

SparkQA commented:
**[Test build #115934 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115934/testReport)** for PR 26805 at commit [`4e1aba9`](https://github.com/apache/spark/commit/4e1aba9c33dceed751db403f303e4bdb0c6e7cf6).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PruneHiveTablePartitions(session: SparkSession)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115934/
Test FAILed.

fuwhu commented:
retest this please

SparkQA commented:
**[Test build #115946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115946/testReport)** for PR 26805 at commit [`4e1aba9`](https://github.com/apache/spark/commit/4e1aba9c33dceed751db403f303e4bdb0c6e7cf6).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20738/
Test PASSed.

SparkQA commented:
**[Test build #115946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115946/testReport)** for PR 26805 at commit [`4e1aba9`](https://github.com/apache/spark/commit/4e1aba9c33dceed751db403f303e4bdb0c6e7cf6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PruneHiveTablePartitions(session: SparkSession)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115946/
Test PASSed.

fuwhu commented:
@cloud-fan @maropu
Could you help review? Or I can close it if this change is not needed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -150,6 +151,100 @@ class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+/**
+ * TODO: merge this with PruneFileSourcePartitions after we completely make hive as a data source.
+ */
+case class PruneHiveTablePartitions(session: SparkSession)
```
let's move it to a separated file.

fuwhu replied:
sure, thanks.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -150,6 +151,100 @@ class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+  /**
+   * Extract the partition filters from the filters on the table.
+   */
+  private def extractPartitionPruningFilters(filters: Seq[Expression],
```
nit:
```
def func(
    para1: T,
    para2: T...
```

fuwhu replied:
sure, thanks.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -150,6 +151,100 @@ class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case op @ PhysicalOperation(projections, filters, relation: HiveTableRelation)
```
2 space indentation.

fuwhu replied:
sure, thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1375,6 +1375,16 @@ object SQLConf {
+      "for statistics calculation is not allowed. This is used to avoid calculating " +
+      "the size of a large number of partitions through hdfs, which is very time consuming." +
+      "Setting this value to 0 or negative will disable falling back to hdfs for " +
+      "partition statistic calculation.")
```
does the data source table have the same problem?

fuwhu replied:
Yes, in PruneFileSourcePartitions, it also may lead to calculating size of large number of partitions through hdfs.
I will create a follow-up PR to refine it.

cloud-fan replied:
If this is a common problem, let's leave it here and open a new PR to fix it completely later.

fuwhu replied:
yea, will create new PR for it. Already removed from this PR. thanks.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -150,6 +151,100 @@ class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+  private def prunedHiveTableWithStats(relation: HiveTableRelation,
+                                       partitionFilters: Seq[Expression]): HiveTableRelation = {
+    val conf = session.sessionState.conf
+    val prunedPartitions = session.sharedState.externalCatalog.listPartitionsByFilter(
```
We should call `SessionCatalog.listPartitionsByFilter`

fuwhu replied:
yea, updated.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -150,6 +151,100 @@ class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+          (part, 0L)
+        }
+      }
+      val sizeOfPartitions =
```
shall we do it in `DetermineTableStats`?

fuwhu replied:
DetermineTableStats is Analyzer rule, and the size of pruned partitions must be calculated after filter push-down optimizers executed. So we can not put this part in DetermineTableStats now.
But I will check whether the DetermineTableStats can be moved to Optimizer and put after PruneHiveTablePartitions. If any idea/suggestion, please share. thanks.

fuwhu committed:
Refine code.

SparkQA commented:
**[Test build #116107 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116107/testReport)** for PR 26805 at commit [`a51e946`](https://github.com/apache/spark/commit/a51e94615e9691b71d9d0f8b219c6d96dd18b599).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20899/
Test PASSed.

SparkQA commented:
**[Test build #116107 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116107/testReport)** for PR 26805 at commit [`a51e946`](https://github.com/apache/spark/commit/a51e94615e9691b71d9d0f8b219c6d96dd18b599).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116107/
Test FAILed.

fuwhu commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20903/
Test PASSed.

SparkQA commented:
**[Test build #116111 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116111/testReport)** for PR 26805 at commit [`a51e946`](https://github.com/apache/spark/commit/a51e94615e9691b71d9d0f8b219c6d96dd18b599).

SparkQA commented:
**[Test build #116111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116111/testReport)** for PR 26805 at commit [`a51e946`](https://github.com/apache/spark/commit/a51e94615e9691b71d9d0f8b219c6d96dd18b599).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116111/
Test PASSed.

fuwhu committed:
Remove conf item FALL_BACK_TO_HDFS_FOR_STATS_MAX_PART_NUM, leaving it to new PR as it is a common problem.

SparkQA commented:
**[Test build #116130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116130/testReport)** for PR 26805 at commit [`d715a04`](https://github.com/apache/spark/commit/d715a04b8e690d068aac3e97f3412a8f477947d2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20922/
Test PASSed.

fuwhu committed:
Refine code.

SparkQA commented:
**[Test build #116141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116141/testReport)** for PR 26805 at commit [`4b8d39d`](https://github.com/apache/spark/commit/4b8d39dcc2affbcf33e61a67d52c6e979cc86402).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20933/
Test PASSed.

SparkQA commented:
**[Test build #116130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116130/testReport)** for PR 26805 at commit [`d715a04`](https://github.com/apache/spark/commit/d715a04b8e690d068aac3e97f3412a8f477947d2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116130/
Test FAILed.

SparkQA commented:
**[Test build #116141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116141/testReport)** for PR 26805 at commit [`4b8d39d`](https://github.com/apache/spark/commit/4b8d39dcc2affbcf33e61a67d52c6e979cc86402).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116141/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,112 @@
+    val newTableMeta =
+      if (relation.tableMeta.stats.isDefined) {
+        relation.tableMeta.copy(
+          stats = Some(relation.tableMeta.stats.get.copy(sizeInBytes = BigInt(sizeInBytes))))
```
is it safe to keep other stats after we prune partitions?

fuwhu replied:
could be inconsistent, eg. rowCount and sizeInBytes is inconsistent.
Restored to new CatalogStatistics instance. But by doing so, some statistics may be lost which should not impact accuracy.

fuwhu committed:
Add PruneHiveTablePartitionsSute.

fuwhu committed:
Refine PruneHiveTablePartitions : prune partitions through metasotre firstly
if HIVE_METASTORE_PARTITION_PRUNING enabled, and then prune again using partition filters.

fuwhu commented:
Did some refining, @cloud-fan please help review again. thanks.

SparkQA commented:
**[Test build #116359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116359/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21151/
Test PASSed.

SparkQA commented:
**[Test build #116359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116359/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116359/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #116367 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116367/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21161/
Test PASSed.

SparkQA commented:
**[Test build #116367 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116367/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116367/
Test FAILed.

fuwhu commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21185/
Test PASSed.

SparkQA commented:
**[Test build #116395 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116395/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).

SparkQA commented:
**[Test build #116395 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116395/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116395/
Test FAILed.

fuwhu commented:
retest this please

SparkQA commented:
**[Test build #116424 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116424/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21213/
Test PASSed.

SparkQA commented:
**[Test build #116424 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116424/testReport)** for PR 26805 at commit [`0218202`](https://github.com/apache/spark/commit/02182020fcfa8d6867d2bedf678b4fdb71ac04ff).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116424/
Test PASSed.

fuwhu commented:
cc @cloud-fan

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,140 @@
+   * Extract the partition filters from the filters on the table.
+   */
+  private def extractPartitionPruningFilters(
+              filters: Seq[Expression],
```
nit: 4 space indentation

fuwhu replied:
sure, thanks.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,140 @@
+  private def extractPartitionPruningFilters(
+              filters: Seq[Expression],
+              relation: HiveTableRelation): Seq[Expression] = {
+    val normalizedFilters = filters.map { e =>
```
let's follow `PruneFileSourcePartitions.getPartitionKeyFilters`

fuwhu replied:
sure, thanks.

cloud-fan replied:
did you follow it? I don't see you filter out non-deterministic filters and other stuff like `PruneFileSourcePartitions.getPartitionKeyFilters` did.

fuwhu replied:
sorry for missing this, updated.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,140 @@
+  /**
+   * Prune the hive table using filters on the partitions of the table.
+   */
+  private def prunePartitions(relation: HiveTableRelation, partitionFilters: Seq[Expression])
```
nit:
```
def f(
    para1: T,
    para2: T): R = ...
```

fuwhu replied:
sure, thanks.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,140 @@
+    } else {
+      session.sessionState.catalog.listPartitions(relation.tableMeta.identifier)
+    }
+    val shouldKeep = partitionFilters.reduceLeftOption(And).map { filter =>
```
do we need this? AFAIK `ExternalCatalog.listPartitionsByFilter` does it already.

fuwhu replied:
Like in HiveTableScanExec, it calls prunePartitions(rawPartitions) in doExecute, while the rawPartitions already call ExternalCatalog.listPartitionsByFilter itself if HIVE_METASTORE_PARTITION_PRUNING enabled. 

It push the partition filters down into hive metastore for earlier pruning, but it still prune it again in prunePartitions, because some predicates may not be supported in hive metastore, like "b like 'xyz'".

So i think it is necessary here, WDYT?

cloud-fan replied:
you are right hive may not support some predicates, so `HiveClientImpl.listPartitionByFilter` may return more partitions than we expect. But it should be fixed in `HiveExternalCatalog.listPartitionsByFilter`

fuwhu replied:
yea, make sense. What about creating a new PR for it and update it in both PruneHiveTablePartitions and HiveTableScanExec after it finished ?

cloud-fan replied:
I mean, it's already fixed, `ExternalCatalog.listPartitionsByFilter` guarantees that the returned partitions are exactly what we want.

fuwhu replied:
yes, it is. Just checked.
Then the HiveTableScanExec should also be updated. right?

cloud-fan replied:
Yea, but we can leave it in followup. The code newly added here should be updated.

fuwhu replied:
yea, updated.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,140 @@
+        } else if (totalSize.isDefined && totalSize.get > 0L) {
+          totalSize.get
+        } else {
+          CommandUtils.calculateLocationSize(
```
It's too expensive to get size for file system for each partition. We should drop the size stats if we can't get it from the partition metadata.

fuwhu replied:
In PruneFileSourcePartitions, it calls "prunedFileIndex.sizeInBytes" which get size from file system for each leaf file (include files of all partitions). I think it's ok to get size from file system for each partition, but we need to limit the number of partitions, just as proposed in #27129 . 
Here, I changed it to drop the size stats if it can't be got from metadata, will update it after #27129 finished.
Does it make sense to you?

cloud-fan replied:
SGTM

fuwhu committed:
Fix indentation.

fuwhu committed:
Drop sizeInBytes of partition when it can't be got from metadata.

SparkQA commented:
**[Test build #116675 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116675/testReport)** for PR 26805 at commit [`3de6491`](https://github.com/apache/spark/commit/3de6491597f2d1b8a011df707d643fba5b14fce8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21455/
Test PASSed.

SparkQA commented:
**[Test build #116675 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116675/testReport)** for PR 26805 at commit [`3de6491`](https://github.com/apache/spark/commit/3de6491597f2d1b8a011df707d643fba5b14fce8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116675/
Test PASSed.

fuwhu committed:
Don't need to prune again in PruneHiveTablePartitions.prunePartitions any more
since HiveExternalCatalog.listPartitionsByFilter can already return exactly what we want.

SparkQA commented:
**[Test build #116700 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116700/testReport)** for PR 26805 at commit [`3e947e8`](https://github.com/apache/spark/commit/3e947e815a3071286aa82d342b0b104f97421fe9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21478/
Test PASSed.

SparkQA commented:
**[Test build #116700 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116700/testReport)** for PR 26805 at commit [`3e947e8`](https://github.com/apache/spark/commit/3e947e815a3071286aa82d342b0b104f97421fe9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116700/
Test PASSed.

fuwhu committed:
Make PruneHiveTablePartitions.getPartitionKeyFilters follows PruneFileSourcePartitions.getPartitionKeyFilters.

SparkQA commented:
**[Test build #116734 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116734/testReport)** for PR 26805 at commit [`45ea071`](https://github.com/apache/spark/commit/45ea07140fe9a2cdb9d85e4e2c2580bf4162719f).

SparkQA commented:
**[Test build #116734 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116734/testReport)** for PR 26805 at commit [`45ea071`](https://github.com/apache/spark/commit/45ea07140fe9a2cdb9d85e4e2c2580bf4162719f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116734/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21509/
Test PASSed.

SparkQA commented:
**[Test build #116736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116736/testReport)** for PR 26805 at commit [`be2c11f`](https://github.com/apache/spark/commit/be2c11f408343b2bee3c0d01a3cb02553764446e).

SparkQA commented:
**[Test build #116736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116736/testReport)** for PR 26805 at commit [`be2c11f`](https://github.com/apache/spark/commit/be2c11f408343b2bee3c0d01a3cb02553764446e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116736/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21511/
Test PASSed.

fuwhu committed:
empty commit

SparkQA commented:
**[Test build #116737 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116737/testReport)** for PR 26805 at commit [`19702f6`](https://github.com/apache/spark/commit/19702f6d919c044913bef74a09f7da3eb17caa99).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21512/
Test PASSed.

SparkQA commented:
**[Test build #116737 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116737/testReport)** for PR 26805 at commit [`19702f6`](https://github.com/apache/spark/commit/19702f6d919c044913bef74a09f7da3eb17caa99).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116737/
Test FAILed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,111 @@
+  private def prunePartitions(
+      relation: HiveTableRelation,
+      partitionFilters: ExpressionSet): Seq[CatalogTablePartition] = {
+    if (conf.metastorePartitionPruning) {
```
instead of check the config here and list all partition if not enabled, we can simply check the config in `apply`:
https://github.com/apache/spark/pull/26805/files#diff-6be42cfa3c62a7536b1eb1d6447c073cR99

fuwhu replied:
what about also pruning it in PruneHiveTablePartitions.prunePartitions even metastorePartitionPruning is not enabled.
Because metastorePartitionPruning should control whether pushing down to hive metastore for early pruning instead of whether do pruning.
Pruning should be done no matter metastorePartitionPruning is enabled or not.
WDYT?

cloud-fan replied:
then can do add a new config for this optimization?

fuwhu replied:
you mean adding a config to control whether we should prune hive table partitions ?

cloud-fan replied:
actually, nvm. Users can exclude optimizer rules to disable them. Let's follow `PruneFileSourcePartitions` and always do pruning.

fuwhu replied:
yes, through "spark.sql.optimizer.excludedRules".
Already updated back.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,111 @@
+        } else if (totalSize.isDefined && totalSize.get > 0L) {
+          totalSize.get
+        } else {
+          0L
```
what if some partitions have size stats and some don't? I think it's safer to not change the table stats if there is one partition that doesn't have size stats.

fuwhu replied:
make sense, will udpate.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,111 @@
+        }
+      }.sum
+    } catch {
+      case e: IOException =>
```
how can we throw IO exception?

fuwhu replied:
yea, it will not now. it may throw if we get size from file system.
will update it.

fuwhu committed:
leave statistic unchanged if the sizeInBytes of some partition is not available in metadata.

SparkQA commented:
**[Test build #116802 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116802/testReport)** for PR 26805 at commit [`0d15c46`](https://github.com/apache/spark/commit/0d15c467481bbd969329e49464f3f4e4e56f7f1e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21575/
Test PASSed.

SparkQA commented:
**[Test build #116802 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116802/testReport)** for PR 26805 at commit [`0d15c46`](https://github.com/apache/spark/commit/0d15c467481bbd969329e49464f3f4e4e56f7f1e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116802/
Test PASSed.

SparkQA commented:
**[Test build #116823 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116823/testReport)** for PR 26805 at commit [`f89947b`](https://github.com/apache/spark/commit/f89947beff6ae5bf61fd83cd646db47381c8db57).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21595/
Test PASSed.

SparkQA commented:
**[Test build #116823 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116823/testReport)** for PR 26805 at commit [`f89947b`](https://github.com/apache/spark/commit/f89947beff6ae5bf61fd83cd646db47381c8db57).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116823/
Test FAILed.

SparkQA commented:
**[Test build #116832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116832/testReport)** for PR 26805 at commit [`acd7fb5`](https://github.com/apache/spark/commit/acd7fb58a9380f580288ab21f9212eae94a3b2d9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21605/
Test PASSed.

fuwhu committed:
refine code.

SparkQA commented:
**[Test build #116843 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116843/testReport)** for PR 26805 at commit [`80b840f`](https://github.com/apache/spark/commit/80b840f66e56362f5abcfb7dbd7e829811f45328).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21615/
Test PASSed.

SparkQA commented:
**[Test build #116832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116832/testReport)** for PR 26805 at commit [`acd7fb5`](https://github.com/apache/spark/commit/acd7fb58a9380f580288ab21f9212eae94a3b2d9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116832/
Test FAILed.

SparkQA commented:
**[Test build #116843 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116843/testReport)** for PR 26805 at commit [`80b840f`](https://github.com/apache/spark/commit/80b840f66e56362f5abcfb7dbd7e829811f45328).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116843/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,109 @@
+      session.sessionState.catalog.listPartitionsByFilter(
+        relation.tableMeta.identifier, partitionFilters.toSeq)
+    } else {
+      ExternalCatalogUtils.prunePartitionsByFilter(relation.tableMeta,
```
can we remove this branch? We should always do the optimization.

fuwhu replied:
you mean totally remove this if clause and always call       "session.sessionState.catalog.listPartitionsByFilter" ?
it means already pushing down to hive metadata for pruning and also make sure returned partitions are exactly what we want. is that expected ?

cloud-fan replied:
sorry I misread the code. so there are 2 optimizations:
1. pushdown predicates to hive metastore
2. pushdown predicates earlier to get precise data size info.

This rule is for the second optimization, and this branch is for skip the first optimization.

makes sense to me

fuwhu replied:
so we don't need change here, right?

cloud-fan replied:
yes

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,109 @@
+        val newTableMeta = updateTableMeta(relation.tableMeta, newPartitions)
+        val newRelation = relation.copy(
+          tableMeta = newTableMeta, prunedPartitions = Some(newPartitions))
+        Project(projections, Filter(filters.reduceLeft(And), newRelation))
```
We shouldn't exclude partition predicates from filters, like what `PruneFileSourcePartitions` does.

fuwhu replied:
PruneFileSourcePartitions keeps the partition filters for LogicalRelation, while exclude partition filters for DataSourceV2ScanRelation. What's the reason of this difference ?

cloud-fan replied:
what do you mean by "keep"?

fuwhu replied:
it pass all filters, including partition filters, to generate new logical plan for LogicalRelation case : 
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala#L86

cloud-fan replied:
because DS v1 needs Spark to carry the pushed filter information, via query plans (so that end-user can see it via EXPLAIN). DS v2 will keep the pushed filter information themselves, so Spark doesn't need to keep it.

fuwhu replied:
By "DS v2 will keep the pushed filter information themselves", you mean the case like  [AvroScan.partitionFilters](https://github.com/apache/spark/blob/master/external/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroScan.scala#L39) ?

cloud-fan replied:
I mean `SupportsPushDownFilters.pushedFilters`. Implementation should carry the pushed filters and report it, then Spark doesn't need to carry it itself.

SparkQA commented:
**[Test build #116987 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116987/testReport)** for PR 26805 at commit [`56249f4`](https://github.com/apache/spark/commit/56249f4d3ec1ca4cf7403dac8a88abedb70b3b00).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21755/
Test PASSed.

fuwhu commented:
@cloud-fan  Already excluded the partition filters, please help review. thanks.

SparkQA commented:
**[Test build #116987 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116987/testReport)** for PR 26805 at commit [`56249f4`](https://github.com/apache/spark/commit/56249f4d3ec1ca4cf7403dac8a88abedb70b3b00).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116987/
Test FAILed.

SparkQA commented:
**[Test build #116992 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116992/testReport)** for PR 26805 at commit [`5b96784`](https://github.com/apache/spark/commit/5b9678455781c3b4083835afb31a525ea530c9f7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21760/
Test PASSed.

SparkQA commented:
**[Test build #116992 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116992/testReport)** for PR 26805 at commit [`5b96784`](https://github.com/apache/spark/commit/5b9678455781c3b4083835afb31a525ea530c9f7).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116992/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21768/
Test PASSed.

SparkQA commented:
**[Test build #117001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117001/testReport)** for PR 26805 at commit [`f97fd69`](https://github.com/apache/spark/commit/f97fd69c2773ec173f1588917155bdccacde6fc5).

SparkQA commented:
**[Test build #117001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117001/testReport)** for PR 26805 at commit [`f97fd69`](https://github.com/apache/spark/commit/f97fd69c2773ec173f1588917155bdccacde6fc5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117001/
Test FAILed.

SparkQA commented:
**[Test build #117073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117073/testReport)** for PR 26805 at commit [`e8ad012`](https://github.com/apache/spark/commit/e8ad012462baac7344a329c94e6ff3b848c76710).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21838/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,118 @@
+          tableMeta = newTableMeta,
+          pushedDownPartitionFilters = Some(partitionKeyFilters.toSeq),
+          prunedPartitions = Some(newPartitions))
+        // The pushed down partition filters don't need to be reevaluated.
```
sorry I was totally wrong. Hive is still a v1 source and we should follow the v1 file source to keep the partitions filters, so that they are visible in physical planning.

Can we change it back? sorry for the back and forth!

fuwhu replied:
sure, will do.

fuwhu replied:
already changed back.

SparkQA commented:
**[Test build #117092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117092/testReport)** for PR 26805 at commit [`23d6d03`](https://github.com/apache/spark/commit/23d6d03f00cdb646da077c845ccaed99df7ccd81).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21856/
Test PASSed.

SparkQA commented:
**[Test build #117073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117073/testReport)** for PR 26805 at commit [`e8ad012`](https://github.com/apache/spark/commit/e8ad012462baac7344a329c94e6ff3b848c76710).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117073/
Test FAILed.

SparkQA commented:
**[Test build #117092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117092/testReport)** for PR 26805 at commit [`23d6d03`](https://github.com/apache/spark/commit/23d6d03f00cdb646da077c845ccaed99df7ccd81).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117092/
Test FAILed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,109 @@
+        val newTableMeta = updateTableMeta(relation.tableMeta, newPartitions)
+        val newRelation = relation.copy(
+          tableMeta = newTableMeta, prunedPartitions = Some(newPartitions))
+        Project(projections, Filter(filters.reduceLeft(And), newRelation))
```
can we add a comment like the file source rule to explain why we keep the partition filter?

fuwhu replied:
sure, thanks.

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala:
```diff
@@ -0,0 +1,61 @@
+      withTempDir { dir =>
+        sql(
+          s"""
+             |CREATE EXTERNAL TABLE test(i int)
```
does it have to be an external table?

fuwhu replied:
actually not necessary, already changed to managed table.

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala:
```diff
@@ -0,0 +1,61 @@
+        val singlePartitionSizeInBytes = 3890
+        val analyzed1 = sql("select i from test where p>0").queryExecution.analyzed
+        val analyzed2 = sql("select i from test where p=1").queryExecution.analyzed
+        assert(Optimize.execute(analyzed1).stats.sizeInBytes === singlePartitionSizeInBytes*4*12/16)
```
checking the accurate file size can be flaky. Can we just check that the first size is more than 3 times larger than the second size?

fuwhu replied:
sure, will change it.

fuwhu committed:
Refine code.

SparkQA commented:
**[Test build #117147 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117147/testReport)** for PR 26805 at commit [`ce20439`](https://github.com/apache/spark/commit/ce204396ed16aecc85f52f4251becb42e1e0c67c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21911/
Test PASSed.

SparkQA commented:
**[Test build #117147 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117147/testReport)** for PR 26805 at commit [`ce20439`](https://github.com/apache/spark/commit/ce204396ed16aecc85f52f4251becb42e1e0c67c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117147/
Test PASSed.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,110 @@
+        0L
+      }
+    }
+    if (sizeOfPartitions.forall(s => s>0)) {
```
nit `forall(_ > 0)`

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala:
```diff
@@ -0,0 +1,56 @@
+
+      for (part <- Seq(1, 2, 3, 4)) {
+        sql(s"""
+               |INSERT OVERWRITE TABLE test PARTITION (p='$part')
```
please keep the style of multiline string consistent. https://github.com/apache/spark/pull/26805/files#diff-90836f7778a704901d7d3df02846aa55R39 is corrected.

fuwhu replied:
updated to use two-space indentation like PruneFileSourcePartitionsSuite.

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala:
```diff
@@ -0,0 +1,56 @@
+               |INSERT OVERWRITE TABLE test PARTITION (p='$part')
+               |select col from temp""".stripMargin)
+      }
+      val analyzed1 = sql("select i from test where p>0").queryExecution.analyzed
```
nit: `p > 0`

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala:
```diff
@@ -0,0 +1,56 @@
+               |select col from temp""".stripMargin)
+      }
+      val analyzed1 = sql("select i from test where p>0").queryExecution.analyzed
+      val analyzed2 = sql("select i from test where p=1").queryExecution.analyzed
```
ditto

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala:
```diff
@@ -0,0 +1,56 @@
+      }
+      val analyzed1 = sql("select i from test where p>0").queryExecution.analyzed
+      val analyzed2 = sql("select i from test where p=1").queryExecution.analyzed
+      assert(Optimize.execute(analyzed1).stats.sizeInBytes/4 ===
```
ditto

fuwhu replied:
updated the code style, thanks a lot. :)

cloud-fan reviewed (approved):
LGTM except a few code style issues.

fuwhu committed:
Fix scala code style.

SparkQA commented:
**[Test build #117166 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117166/testReport)** for PR 26805 at commit [`b1798d5`](https://github.com/apache/spark/commit/b1798d52147b081c7073f3c096eb886a867b921d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21930/
Test PASSed.

SparkQA commented:
**[Test build #117166 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117166/testReport)** for PR 26805 at commit [`b1798d5`](https://github.com/apache/spark/commit/b1798d52147b081c7073f3c096eb886a867b921d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117166/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

fuwhu commented:
Thank you all for review and help.

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala:
```diff
@@ -0,0 +1,110 @@
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * TODO: merge this with PruneFileSourcePartitions after we completely make hive as a data source.
```
@fuwhu We need a description about the rule. Could you submit a follow-up PR to add the descriptions to both PruneHiveTablePartitions and PruneFileSourcePartitions?

fuwhu replied:
sure, so you mean just add class description in PruneHiveTablePartitions.scala and PruneFileSourcePartitions.scala file ? Or need to add comment in some doc ?

cloud-fan replied:
classdoc is good enough

fuwhu replied:
@gatorsmile @cloud-fan classdoc added in #27535 , please help review, thanks.
